api/lib/flight_job_script_api/system_command.rb

```diff
@@ -0,0 +1,202 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Script Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Script Service, please visit:
+# https://github.com/openflighthpc/flight-job-script-service
+#==============================================================================
+
+module FlightJobScriptAPI
+  class CommandError < Sinja::ServiceUnavailable; end
+
+  class SystemCommand
+    # Used to ensure each user is only running a single command at at time
+    # NOTE: These objects will be indefinitely cached in memory until the server
+    #       is restarted. This may constitute a memory leak if an indefinite
+    #       number of users access the service.
+    #       Consider refactoring
+    def self.mutexes
+      @mutexes ||= Hash.new { |h, k| h[k] = Mutex.new }
+    end
+
+    def self.flight_list_templates(**opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'list-templates', '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_info_template(id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'info-template', id, '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_list_scripts(**opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'list-scripts', '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_info_script(id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'info-script', id, '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_delete_script(id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'delete-script', id, '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_create_script(template_id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'create-script', template_id, '--json', '--stdin', **opts).tap(&:wait)
+    end
+
+    def self.flight_list_jobs(**opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'list-jobs', '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_info_job(id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'info-job', id, '--json', **opts).tap(&:wait)
+    end
+
+    def self.flight_submit_job(script_id, **opts)
+      new(*FlightJobScriptAPI.config.flight_job, 'submit-job', script_id, '--json', **opts).tap(&:wait)
+    end
+
+    attr_reader :cmd, :user, :mutex, :stdin
+    attr_accessor :stdout, :stderr, :exitstatus
+
+    def initialize(*cmd, user:, mutex: nil, stdin: nil)
+      @cmd = cmd
+      @user = user
+      @stdin = stdin
+      @mutex = self.class.mutexes[user]
+    end
+
+    def wait
+      return if @wait
+      run
+      @wait ||= true
```

Comment:
This looks ood.  What could `@wait` be set to here that doesn't result in it being set to `true`?

Reply:
We don't have any form of promises in the `SystemCommand` class, so the `wait` method takes its place. Its purposes is to ensure the system command has been ran before returning, but also ensures it runs only once.

The actual value of `@wait` doesn't really matter, apart from the fact it is truthy.

Reply:
It's the use of `||=` rather than `=` that confuses me.  It suggests that `@wait` could be set to something truthy inside of `run` and that we want to preserve that value.

Reply:
b51cc0e265f8829ae2aafa56ff0f1975eca4953d